kbase-extension/static/kbase/js/widgets/narrative_core/upload/importErrors.js

```diff
@@ -0,0 +1,246 @@
+                    this.fileErrors[fileKey] = [];
+                }
+                this.fileErrors[fileKey].push(error.message);
+            };
+
+            this.errors.forEach((error) => {
+                switch (error.type) {
+                    case BULK_SPEC_ERRORS.CANNOT_PARSE:
+                    case BULK_SPEC_ERRORS.INCORRECT_COLUMN_COUNT:
+                        addFileError(error);
+                        break;
+                    case BULK_SPEC_ERRORS.NOT_FOUND:
+                        addFileError(Object.assign({ message: 'File not found' }, error));
+                        break;
+                    case BULK_SPEC_ERRORS.MULTIPLE_SPECS:
+                        addFileError({
+                            file: error.file_1,
+                            tab: error.tab_1,
+                            message: error.message,
+                        });
+                        addFileError({
+                            file: error.file_2,
+                            tab: error.tab_2,
+                            message: error.message,
+                        });
+                        break;
+                    case BULK_SPEC_ERRORS.NO_FILES:
+                        this.noFileError = true;
+                        break;
+                    case BULK_SPEC_ERRORS.UNKNOWN:
+                        if (error.file) {
+                            addFileError(error);
+                        } else {
+                            this.unexpectedErrors.push(
+                                error.message || 'An unexpected error occurred!'
+                            );
+                        }
+                        break;
+                    case BULK_SPEC_ERRORS.SERVER:
+                        this.serverErrors.push(error.message);
+                        break;
+                    default:
+                        if (error.message) {
+                            this.unexpectedErrors.push(error.message);
+                        } else if (error.type) {
+                            this.unexpectedErrors.push(`Unknown error of type "${error.type}"`);
+                        } else {
+                            this.unexpectedErrors.push(`An unknown error occurred!`);
+                        }
+                        console.error('Unexpected import setup error!', error);
+                        break;
+                }
+            });
+        }
+
+        _formatErrors() {
+            const tag = html.tag,
+                div = tag('div'),
+                ul = tag('ul'),
+                li = tag('li'),
+                b = tag('b');
+
+            let title = 'Bulk import error';
+            let body = '';
+            // some branching based on how errors were parsed out.
+
+            if (this.noFileError) {
+                title = 'No files provided';
+                body =
+                    'No CSV/TSV/Excel files were provided, but "Import Specification" was selected.';
```

Comment:
You could return this and the next option straight away since the only shared code between the all three options in the if/else is just `return [title, body]` -- it'd decrease the complexity a little and make the function easier to parse. I'd also be tempted to return the results as `{ title, body }` for extra clarity, but that's more of a personal style choice.

Reply:
Done! `{ title, body }` is much more clear, too, thanks.